packages/ethereum-contracts/contracts/apps/SuperAppBaseFlow.sol

```diff
@@ -0,0 +1,265 @@
+        revert("Unsupported callback - Before Agreement Deleted");
+    }
+
+    // ---------------------------------------------------------------------------------------------
+    // AFTER- CALLBACKS
+
+    function afterAgreementCreated(
+        ISuperToken superToken,
+        address agreementClass,
+        bytes32 /*agreementId*/,
+        bytes calldata agreementData,
+        bytes calldata cbdata,
+        bytes calldata ctx
+    ) external override returns (bytes memory newCtx) {
+        // sender is host
+        if (msg.sender != address(_host)) revert UnauthorizedHost();
+        // super token is accepted
+        if (!_isAcceptedSuperToken(superToken)) return ctx;
+        // agreement is CFA
+        if (!_isAcceptedAgreement(agreementClass)) return ctx;
+
+        (address sender, ) = abi.decode(agreementData, (address, address));
+
+        return
+            afterFlowCreated(
+                superToken,
+                sender,
+                cbdata,
+                ctx // userData can be acquired with `_host.decodeCtx(ctx).userData`
+            );
+    }
+
+    function afterAgreementUpdated(
+        ISuperToken superToken,
+        address agreementClass,
+        bytes32 /*agreementId*/,
+        bytes calldata agreementData,
+        bytes calldata cbdata,
+        bytes calldata ctx
+    ) external override returns (bytes memory newCtx) {
+        // sender is host
+        if (msg.sender != address(_host)) revert UnauthorizedHost();
+        // super token is accepted
+        if (!_isAcceptedSuperToken(superToken)) return ctx;
+        // agreement is CFA
+        if (!_isAcceptedAgreement(agreementClass)) return ctx;
+
+        (address sender, ) = abi.decode(agreementData, (address, address));
+
+        return
+            afterFlowUpdated(
+                superToken,
+                sender,
+                cbdata,
+                ctx // userData can be acquired with `_host.decodeCtx(ctx).userData`
+            );
+    }
+
+    function afterAgreementTerminated(
+        ISuperToken superToken,
+        address agreementClass,
+        bytes32 /*agreementId*/,
+        bytes calldata agreementData,
+        bytes calldata cbdata,
+        bytes calldata ctx
+    ) external override returns (bytes memory newCtx) {
+        // sender is host
+        if (msg.sender != address(_host)) return ctx;
+        // super token is accepted
+        if (!_isAcceptedSuperToken(superToken)) return ctx;
+        // agreement is CFA
+        if (!_isAcceptedAgreement(agreementClass)) return ctx;
+
+        (address sender, address receiver) = abi.decode(agreementData, (address, address));
+
+        return 
+            afterFlowDeleted(
+                superToken,
+                sender,
+                receiver,
+                cbdata,
+                ctx
+            );
+    }
+
+    // ---------------------------------------------------------------------------------------------
+    // BEFORE- CALLBACKS
```

Comment:
maybe place those before the "after" callbacks?

Reply:
I put the "before" callbacks after the "after" callbacks (lol) as the "after" callbacks are of more interest to the developer and are what's actually overriden and used. The other stuff is behind the scenes. But I can switch them around if you think we should be more conventional